File: core/src/main/java/com/ustadmobile/core/controller/SelectMultipleProductTypeTreeDialogPresenter.java
package com.ustadmobile.core.controller;

import com.ustadmobile.core.db.UmAppDatabase;
import com.ustadmobile.core.db.dao.SaleProductParentJoinDao;
import com.ustadmobile.core.impl.UmAccountManager;
import com.ustadmobile.core.impl.UmCallback;
import com.ustadmobile.core.view.SelectMultipleProductTypeTreeDialogView;
import com.ustadmobile.lib.db.entities.SaleProduct;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Hashtable;
import java.util.List;

import static com.ustadmobile.core.controller.ReportOptionsDetailPresenter.convertCSVStringToLongList;
import static com.ustadmobile.core.view.SelectMultipleProductTypeTreeDialogView.ARG_PRODUCT_SELECTED_SET;

/**
 * The SelectMultipleTreeDialog Presenter.
 */
public class SelectMultipleProductTypeTreeDialogPresenter
        extends CommonEntityHandlerPresenter<SelectMultipleProductTypeTreeDialogView> {

    private HashMap<String, Long> selectedOptions;

    private List<Long> selectedProductTypeUidsList;

    UmAppDatabase repository = UmAccountManager.getRepositoryForActiveAccount(context);

    public SelectMultipleProductTypeTreeDialogPresenter(Object context, Hashtable arguments,
                                                        SelectMultipleProductTypeTreeDialogView view) {
        super(context, arguments, view);

        if (arguments.containsKey(ARG_PRODUCT_SELECTED_SET)) {
            String productTypesArrayString = arguments.get(ARG_PRODUCT_SELECTED_SET).toString();
            selectedProductTypeUidsList = convertCSVStringToLongList(productTypesArrayString);
        }

        selectedOptions = new HashMap<>();
        getTopProductTypes();

    }

    static ArrayList<Long> convertLongArray(long[] array) {
        ArrayList<Long> result = new ArrayList<Long>(array.length);
        for (long item : array)
            result.add(item);
        return result;
    }

    public HashMap<String, Long> getSelectedOptions() {
        return selectedOptions;
    }

    private void getTopProductTypes() {
        SaleProductParentJoinDao parentJoinDao = repository.getSaleProductParentJoinDao();
        parentJoinDao.findTopSaleProductsAsync(new UmCallback<List<SaleProduct>>() {
            @Override
            public void onSuccess(List<SaleProduct> result) {
                view.populateTopProductType(result);
            }

            @Override
            public void onFailure(Throwable exception) { exception.printStackTrace(); }
        });
    }

    @Override
    public void onCreate(Hashtable savedState) {
        super.onCreate(savedState);
    }


    public void handleClickPrimaryActionButton() {
        view.finish();
    }


    public List<Long> getSelectedProductTypeUidsList() {
        if (selectedProductTypeUidsList == null) {
            return new ArrayList<>();
        }
        return selectedProductTypeUidsList;
    }

    @Override
    public void entityChecked(String entityName, Long entityUid, boolean checked) {
        if (checked) {
            selectedOptions.put(entityName, entityUid);
        } else {
            selectedOptions.remove(entityName);
        }
    }

}

File: core/src/commonMain/kotlin/com/ustadmobile/core/controller/SelectMultipleProductTypeTreeDialogPresenter.kt
package com.ustadmobile.core.controller

import com.ustadmobile.core.impl.UmAccountManager
import com.ustadmobile.core.util.ext.observeWithPresenter
import com.ustadmobile.core.view.SelectMultipleProductTypeTreeDialogView
import com.ustadmobile.core.view.SelectMultipleProductTypeTreeDialogView.Companion.ARG_PRODUCT_SELECTED_SET
import com.ustadmobile.lib.db.entities.SaleProduct
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.Runnable
import kotlinx.coroutines.launch


/**
 * The SelectMultipleTreeDialog Presenter.
 */
class SelectMultipleProductTypeTreeDialogPresenter(context: Any, arguments:
                            Map<String, String>?,view: SelectMultipleProductTypeTreeDialogView)
    : CommonEntityHandlerPresenter<SelectMultipleProductTypeTreeDialogView>
        (context, arguments!!, view) {

    val selectedOptions: HashMap<String, Long>

    private var selectedProductTypeUidsList: List<Long>? = null

    internal var repository = UmAccountManager.getRepositoryForActiveAccount(context)

    init {

        if (arguments!!.containsKey(ARG_PRODUCT_SELECTED_SET)) {
            val productTypesArrayString = arguments!!.get(ARG_PRODUCT_SELECTED_SET).toString()
            selectedProductTypeUidsList = convertCSVStringToLongList(productTypesArrayString)
        }

        selectedOptions = HashMap()
        getTopProductTypes()

    }

    fun convertCSVStringToLongList(csString:String):List<Long> {
        val list = ArrayList<Long>()
        for (s in csString.split((",").
                toRegex()).dropLastWhile({ it.isEmpty() }).toTypedArray()){
            val p = s.trim()
            list.add(p as Long)
        }

        return list
    }

    private fun getTopProductTypes() {
        val parentJoinDao = repository.saleProductParentJoinDao
        val thisP=this
        GlobalScope.launch {
            val resultLive = parentJoinDao.findTopSaleProductsLive()
            GlobalScope.launch(Dispatchers.Main) {
                resultLive.observeWithPresenter(thisP, thisP::handleProductTypes)
            }
        }
    }

    private fun handleProductTypes(types:List<SaleProduct>?){
        view.runOnUiThread(Runnable {
            view.populateTopProductType(types!!)
        })
    }

    fun handleClickPrimaryActionButton() {
        view.finish()
    }


    fun getSelectedProductTypeUidsList(): List<Long> {

        if (selectedProductTypeUidsList == null){
            return ArrayList()
        }

        return selectedProductTypeUidsList as List<Long>


    }

    override fun entityChecked(entityName: String, entityUid: Long, checked: Boolean) {
        if (checked) {
            if (entityUid != null) {
                selectedOptions.put(entityName, entityUid)
            }
        } else {
            selectedOptions.remove(entityName)
        }
    }

    companion object {

        internal fun convertLongArray(array: LongArray): ArrayList<Long> {
            val result = ArrayList<Long>(array.size)
            for (item in array)
                result.add(item)
            return result
        }
    }

}
